File: TripKitUI/helper/Categories/UIView+BearingRotation.m
//
//  UIView+BearingRotation.m
//  TripPlanner
//
//  Created by Adrian Schoenig on 20/11/12.
//
//

#import "UIView+BearingRotation.h"

@implementation UIView (BearingRotation)

+ (CGFloat)bearingToRotation:(CGFloat)bearing
{
  // 0 = North, 90 = East, 180 = South and 270 = West
  CGFloat start = 90.0f;
  CGFloat rotation = -1 * (start - bearing);
  return (rotation * (CGFloat)M_PI) / 180.0f;
}

- (void)rotateForBearing:(CGFloat)bearing
{
	CGFloat rotation = [[self class] bearingToRotation:bearing];
	self.transform = CGAffineTransformMakeRotation(rotation);
}

- (void)updateForMagneticHeading:(CGFloat)heading andBearing:(CGFloat)bearing
{
	[self rotateForBearing:bearing - heading];
}

@end

File: TripKitUI/helper/Categories/UIView+BearingRotation.swift
//
//  UIView+BearingRotation.swift
//  TripKitUI-iOS
//
//  Created by Adrian Schönig on 11/5/21.
//  Copyright © 2021 SkedGo Pty Ltd. All rights reserved.
//

import UIKit

extension UIView {
  func update(magneticHeading: CGFloat, bearing: CGFloat) {
    rotate(bearing: bearing - magneticHeading)
  }
  
  func rotate(bearing: CGFloat) {
    let rotation = Self.rotating(from: bearing)
    transform = CGAffineTransform(rotationAngle: rotation)
  }
  
  private static func rotating(from bearing: CGFloat) -> CGFloat {
    // 0 = North, 90 = East, 180 = South and 270 = West
    let start: CGFloat = 90
    let rotation = -1 * (start - bearing)
    return (rotation * CGFloat.pi) / 180
  }
}
